streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java

```diff
@@ -187,225 +235,252 @@ public Subscription subscription(Set<String> topics) {
+                int numPartitions = -1;
+                for (TaskId task : tasksByTopicGroup.get(topicGroupId)) {
+                    if (numPartitions < task.partition + 1)
+                        numPartitions = task.partition + 1;
                 }
 
-                tasks.add(task);
+                InternalTopicMetadata topicMetadata = new InternalTopicMetadata(topicConfig);
+                topicMetadata.numPartitions = numPartitions;
+
+                changelogTopicMetadata.put(topicConfig.name(), topicMetadata);
             }
         }
 
+        prepareTopic(changelogTopicMetadata);
+
+        log.debug("stream-thread [{}] Created state changelog topics {} from the parsed topology.", streamThread.getName(), changelogTopicMetadata);
+
+        // ---------------- Step Two ---------------- //
+
         // assign tasks to clients
-        states = TaskAssignor.assign(states, partitionsForTask.keySet(), numStandbyReplicas, streamThread.getName());
+        Map<UUID, ClientState<TaskId>> states = new HashMap<>();
+        for (Map.Entry<UUID, ClientMetadata> entry : clientsMetadata.entrySet()) {
+            states.put(entry.getKey(), entry.getValue().state);
+        }
+
+        log.debug("stream-thread [{}] Assigning tasks {} to clients {} with number of replicas {}",
+                streamThread.getName(), partitionsForTask.keySet(), states, numStandbyReplicas);
 
-        final List<AssignmentSupplier> assignmentSuppliers = new ArrayList<>();
+        TaskAssignor.assign(states, partitionsForTask.keySet(), numStandbyReplicas);
 
-        final Map<HostInfo, Set<TopicPartition>> endPointMap = new HashMap<>();
-        for (Map.Entry<UUID, Set<String>> entry : consumersByClient.entrySet()) {
-            UUID processId = entry.getKey();
-            Set<String> consumers = entry.getValue();
-            ClientState<TaskId> state = states.get(processId);
+        log.info("stream-thread [{}] Assigned tasks to clients as {}.", streamThread.getName(), states);
+
+        // ---------------- Step Three ---------------- //
+
+        // construct the global partition assignment per host map
+        partitionsByHostState = new HashMap<>();
+        for (Map.Entry<UUID, ClientMetadata> entry : clientsMetadata.entrySet()) {
+            ClientState<TaskId> state = entry.getValue().state;
+            final Set<TopicPartition> topicPartitions = new HashSet<>();
+
+            for (TaskId id : state.activeTasks) {
+                for (TopicPartition partition : partitionsForTask.get(id)) {
+                    topicPartitions.add(partition);
+                }
+            }
+
+            for (TaskId id : state.assignedTasks) {
+                for (TopicPartition partition : partitionsForTask.get(id)) {
+                    topicPartitions.add(partition);
+                }
+            }
+
+            HostInfo hostInfo = entry.getValue().hostInfo;
+
+            if (hostInfo != null) {
```

Comment:
Should we move this to the beginning of the loop? If we do not add anything to `partitionsByHostState` there is no reason to create `topicPartitions` in the first place. Or do I miss anything?